src/proof.js

```diff
@@ -13,112 +14,74 @@ export const path = ([, path]) => path
+ * @returns {API.Result<{}, Error>}
  */
-export function validateLeaf(data, root, proof) {
-  const leaf = truncatedHash(data)
-  return validateSubtree(leaf, root, proof)
-}
+export const verify = (proof, { tree, node }) => {
+  const computedRoot = resolveRoot(proof, node)
+  if (computedRoot.error) {
+    return { error: new Error(`computing root: ${computedRoot.error.message}`) }
+  }
 
-/**
- * @param {API.MerkleTreeNode} subtree
- * @param {API.MerkleTreeNode} root
- * @param {API.ProofData} proof
- * @returns {API.Result<void, Error>}
- */
-export function validateSubtree(subtree, root, proof) {
-  // Validate the structure first to avoid panics
-  const structureValidation = validateProofStructure(proof)
-  if (structureValidation.error) {
+  if (!Bytes.equals(computedRoot.ok, tree)) {
     return {
-      error: new Error(
-        `in ValidateSubtree: ${structureValidation.error.message}`
-      ),
+      error: new Error('inclusion proof does not lead to the same root'),
     }
   }
-  return validateProof(subtree, root, proof)
+  return { ok: {} }
 }
 
 const MAX_DEPTH = 63
 
 /**
- * @param {API.MerkleTreeNode} subtree
- * @param {API.ProofData} proofData
- * @returns {API.Result<API.MerkleTreeNode, Error>}
+ * Resolves the root of the merkle tree from given proof and node that root
+ * supposedly includes. It does so by computing parent node from provided node
+ * and node in the proof path, then combining that with the next node in the
+ * path and so on until the root is reached. Function may return an error if
+ * proof path is too long or if proof offset falls out of bounds.
+ *
+ * @param {API.ProofData} proof
+ * @param {API.MerkleTreeNode} node
+ * @returns {API.Result<API.MerkleTreeNode, RangeError>}
  */
-export function computeRoot(subtree, proofData) {
-  if (subtree === null) {
-    return { error: new Error('nil subtree cannot be used') }
-  }
-  if (depth(proofData) > MAX_DEPTH) {
+export function resolveRoot(proof, node) {
+  if (depth(proof) > MAX_DEPTH) {
     return {
-      error: new Error(
-        'merkleproofs with depths greater than 63 are not supported'
+      error: new RangeError(
+        'merkle proofs with depths greater than 63 are not supported'
       ),
     }
   }
 
-  let index = at(proofData)
-  if (index >> BigInt(depth(proofData)) !== 0n) {
-    return { error: new Error('index greater than width of the tree') }
+  let position = offset(proof)
+  if (position >> BigInt(depth(proof)) !== 0n) {
+    return { error: new RangeError('offset greater than width of the tree') }
   }
 
-  let carry = subtree
+  let top = node
   let right = 0n
 
-  for (const p of path(proofData)) {
-    ;[right, index] = [index & 1n, index >> 1n]
-    carry = right === 1n ? computeNode(p, carry) : computeNode(carry, p)
+  for (const node of path(proof)) {
+    ;[right, position] = [position & 1n, position >> 1n]
+    top = right === 1n ? computeNode(node, top) : computeNode(top, node)
```

Comment:
this library is so full of clever, i think maybe we can be simple about assigning variables. Let lets be lets
```suggestion
  for (const node of path(proof)) {
    right =  position & 1n
    position = position >> 1n
    top = right === 1n ? computeNode(node, top) : computeNode(top, node)
```